### PR TITLE
Add support for {1D,2D,3D} arrays in shader compiler and migrate math ops

### PR DESCRIPTION
**Pull request by dsmilkov**
Also migrate to logical sampling
 - unary_op and its children (relu,log,exp,tanh,...)
 - reduce_sum
 - logsumexp
 - min/max
 - arg min/max
 - argmaxequals

This is related to #9 and #18

---
This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/21)

### CODE REVIEW

File: src/math/webgl/gpgpu_context.ts [hunk 1]
```diff
@@ -166,6 +180,23 @@ export class GPGPUContext {
     return program;
   }
 
+  public compileProgram(program: GPGPUProgram, inputs: NDArray[],
```
Comment (nsthorat) [inline src/math/webgl/gpgpu_context.ts]: I think this is unused, you have a gpgpu_math file

**Comment by nsthorat** [comment]
Reviewed 4 of 40 files at r1.
Review status: 4 of 40 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/math/math_gpu.ts, line 452 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-KrAivXPW1xj4M25iWIZ:-KrAivXPW1xj4M25iWI_:baz84pp) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b3051209fead4bf068a21b3b0f0041b62da3a69c/src/math/math_gpu.ts#L452)):*
> ```XML
>     const key = this.makeShaderKey(program, [a], out);
>     const binary = this.getAndSaveCompiledProgram(key, () => {
>       return gpgpu_math.compileProgram(this.gpgpu, program, [a], out);
> ```

what about if you made the shader key part of the binary? and when you call gpgu_math.runProgram, thats when the shader program gets compiled. When you construct the ShaderProgram (ReduceSumProgram in this case) you would pass it [a], and out, so you wouldn't have to do that twice (then the code cleans up really really nicely):

the api would look something as simple as two lines of code:

protected sumInternal(a: NDArray): Scalar {
  const program = new ReduceSumProgram([a]);
  return this.executeProgram(program);
}

// internal implementation of math.executeProgram
executeProgram(program: GPGPUProgram) {
  const key = program.getKey();
  const binary = this.getAndSaveCompiledProgram(key, () => {
    return gpgpu_math.compileProgram(program); // program has all the information about inputs and outputs, so gpgpu_math can call getInputs, getOutputs, etc
  });
  return gpgpu_math.runProgram(binary, program); // this now returns the NDArray. The program itself has all of the information about output shape, the binary is the actual executable that might have been cached, so it can allocate the underlying texture by calling a method on program (so we don't do this stuff in two places).
}

If you want to add a new shader, literally all you have to do is implement these couple lines of code (and not duplicate all the caching stuff all over the place), and implement a shader program in its own file has a method that simply allocates an output NDArray. All of the logic for that program now lives in a single place, and not two!

What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21)*

**Comment by dsmilkov** [comment]
Reviewed 31 of 40 files at r1, 5 of 5 files at r2.
Review status: 34 of 35 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/math/math_gpu.ts, line 452 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-KrAivXPW1xj4M25iWIZ:-KrBZsFQ3tPmH7VV7J_8:bv03hwe) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b3051209fead4bf068a21b3b0f0041b62da3a69c/src/math/math_gpu.ts#L452)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

what about if you made the shader key part of the binary? and when you call gpgu_math.runProgram, thats when the shader program gets compiled. When you construct the ShaderProgram (ReduceSumProgram in this case) you would pass it [a], and out, so you wouldn't have to do that twice (then the code cleans up really really nicely):

the api would look something as simple as two lines of code:

protected sumInternal(a: NDArray): Scalar {
  const program = new ReduceSumProgram([a]);
  return this.executeProgram(program);
}

// internal implementation of math.executeProgram
executeProgram(program: GPGPUProgram) {
  const key = program.getKey();
  const binary = this.getAndSaveCompiledProgram(key, () => {
    return gpgpu_math.compileProgram(program); // program has all the information about inputs and outputs, so gpgpu_math can call getInputs, getOutputs, etc
  });
  return gpgpu_math.runProgram(binary, program); // this now returns the NDArray. The program itself has all of the information about output shape, the binary is the actual executable that might have been cached, so it can allocate the underlying texture by calling a method on program (so we don't do this stuff in two places).
}

If you want to add a new shader, literally all you have to do is implement these couple lines of code (and not duplicate all the caching stuff all over the place), and implement a shader program in its own file has a method that simply allocates an output NDArray. All of the logic for that program now lives in a single place, and not two!

What do you think?
</blockquote></details>

+1 to most points here. The only thing I'll push back is to not have the gpgpuprogram allocate the result ndarray itself, since we have shaders like copy which don't need to allocate a new texture. Also it decouples the way we allocate next textures (texture manager) from the shader logical.

---

*[src/math/webgl/gpgpu_context.ts, line 183 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-KrAfo4yUjcl_Fn6FrqV-r1-183:-KrBRPmkhF-6SSIEBd9g:biokg0t) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b3051209fead4bf068a21b3b0f0041b62da3a69c/src/math/webgl/gpgpu_context.ts#L183)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I think this is unused, you have a gpgpu_math file
</blockquote></details>

Oh my, gpgpu_context.ts was supposed to be untouched. I started here and moved everything to gpgpu_math. I remember deleting all of this - but of course git says otherwise :) - will revert the whole file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21)*

**Comment by dsmilkov** [comment]
Review status: 34 of 35 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/math/math_gpu.ts, line 452 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-KrAivXPW1xj4M25iWIZ:-KrBbb4ueE2qbfrhXeQ_:b-mios8r) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b3051209fead4bf068a21b3b0f0041b62da3a69c/src/math/math_gpu.ts#L452)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

+1 to most points here. The only thing I'll push back is to not have the gpgpuprogram allocate the result ndarray itself, since we have shaders like copy which don't need to allocate a new texture. Also it decouples the way we allocate next textures (texture manager) from the shader logical.
</blockquote></details>

One more thing: I'm not 100% sure about having runProgram(binary, program). I like runProgram(binary, args) since you compile it once and run it many times with different input/output arguments, without the need to construct new Program() every time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21)*

**Comment by dsmilkov** [comment]
Review status: 17 of 35 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/math/math_gpu.ts, line 452 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-KrAivXPW1xj4M25iWIZ:-KrCHf59K1OgzlYSeItn:bv1qgqz) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b3051209fead4bf068a21b3b0f0041b62da3a69c/src/math/math_gpu.ts#L452)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

One more thing: I'm not 100% sure about having runProgram(binary, program). I like runProgram(binary, args) since you compile it once and run it many times with different input/output arguments, without the need to construct new Program() every time.
</blockquote></details>

Went with runProgram(binary, program) for simplicity. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21)*

**Comment by nsthorat** [comment]
Reviewed 1 of 40 files at r1, 1 of 18 files at r3, 17 of 17 files at r4.
Review status: 34 of 35 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/math/webgl/shader_compiler.ts, line 316 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-KrDgS9hOU4aZmtoe6q0:-KrDgS9hOU4aZmtoe6q1:bqx6ynr) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b9c5f7a6bab865c04f7327ae529da55ecee815ab/src/math/webgl/shader_compiler.ts#L316)):*
> ```XML
>     }
>   `;
> ```

remove blank line

---

*[src/math/webgl/unaryop_gpu.ts, line 70 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-KrDgkMJohbrLKT6GPKa:-KrDgkMKFS3CcAd8nAe_:bix86vq) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b9c5f7a6bab865c04f7327ae529da55ecee815ab/src/math/webgl/unaryop_gpu.ts#L70)):*
> ```XML
> }
> 
> export function uploadUnaryDownload(a: NDArray, op: UnaryOp): Float32Array {
> ```

can you move this to a unaryop_gpu_test_util.ts file for consistency (it's not a real test file, it's just a util)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21)*

**Comment by nsthorat** [comment]
:lgtm:

---

Review status: 34 of 35 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-:-KrDh4SOeo0ShiTEePc5:bnfp4nl)*

**Comment by nsthorat** [comment]
Reviewed 1 of 17 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21)*

**Comment by dsmilkov** [comment]
Reviewed 2 of 40 files at r1, 4 of 18 files at r3, 17 of 17 files at r4.
Review status: 25 of 36 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/math/webgl/shader_compiler.ts, line 316 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-KrDgS9hOU4aZmtoe6q0:-KrDhhTb1t3JBUIOzrr4:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b9c5f7a6bab865c04f7327ae529da55ecee815ab/src/math/webgl/shader_compiler.ts#L316)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

remove blank line
</blockquote></details>

Done.

---

*[src/math/webgl/unaryop_gpu.ts, line 70 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-KrDgkMJohbrLKT6GPKa:-KrDjYCgqr0OQZTjftUC:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b9c5f7a6bab865c04f7327ae529da55ecee815ab/src/math/webgl/unaryop_gpu.ts#L70)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you move this to a unaryop_gpu_test_util.ts file for consistency (it's not a real test file, it's just a util)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21)*

**Comment by dsmilkov** [comment]
Reviewed 1 of 1 files at r5, 11 of 11 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21)*

**Comment by dsmilkov** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/math/webgl/unaryop_gpu.ts, line 70 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-KrDgkMJohbrLKT6GPKa:-KrDkB3hMkqoK69mH0Xa:b32w0x) ([raw file](https://github.com/pair-code/deeplearnjs/blob/b9c5f7a6bab865c04f7327ae529da55ecee815ab/src/math/webgl/unaryop_gpu.ts#L70)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21)*

**Comment by nsthorat** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/21#-:-KrDp0ytysXhRiPhbOnP:bnfp4nl)*

**Pull request merged**